extensions/cornerstone/src/Viewport/OHIFCornerstoneViewport.tsx

```diff
@@ -28,428 +29,401 @@ const STACK = 'stack';
-  const nextDisplaySets = nextProps.displaySets;
-
-  if (prevDisplaySets.length !== nextDisplaySets.length) {
-    return false;
-  }
-
-  for (let i = 0; i < prevDisplaySets.length; i++) {
-    const prevDisplaySet = prevDisplaySets[i];
-
-    const foundDisplaySet = nextDisplaySets.find(
-      nextDisplaySet =>
-        nextDisplaySet.displaySetInstanceUID === prevDisplaySet.displaySetInstanceUID
-    );
-
-    if (!foundDisplaySet) {
-      return false;
+// Todo: This should be done with expose of internal API similar to react-vtkjs-viewport
+// Then we don't need to worry about the re-renders if the props change.
+const OHIFCornerstoneViewport = React.memo(
+  (
+    props: withAppTypes<{
+      viewportId: string;
+      displaySets: AppTypes.DisplaySet[];
+      viewportOptions: AppTypes.ViewportGrid.GridViewportOptions;
+      initialImageIndex: number;
+    }>
+  ) => {
+    const {
+      displaySets,
+      dataSource,
+      viewportOptions,
+      displaySetOptions,
+      servicesManager,
+      commandsManager,
+      onElementEnabled,
+      // eslint-disable-next-line react/prop-types
+      onElementDisabled,
+      isJumpToMeasurementDisabled = false,
+      // Note: you SHOULD NOT use the initialImageIdOrIndex for manipulation
+      // of the imageData in the OHIFCornerstoneViewport. This prop is used
+      // to set the initial state of the viewport's first image to render
+      // eslint-disable-next-line react/prop-types
+      initialImageIndex,
+      // if the viewport is part of a hanging protocol layout
+      // we should not really rely on the old synchronizers and
+      // you see below we only rehydrate the synchronizers if the viewport
+      // is not part of the hanging protocol layout. HPs should
+      // define their own synchronizers. Since the synchronizers are
+      // viewportId dependent and
+      // eslint-disable-next-line react/prop-types
+      isHangingProtocolLayout,
+    } = props;
+    const viewportId = viewportOptions.viewportId;
+
+    if (!viewportId) {
+      throw new Error('Viewport ID is required');
     }
 
-    // check they contain the same image
-    if (foundDisplaySet.images?.length !== prevDisplaySet.images?.length) {
-      return false;
+    // Make sure displaySetOptions has one object per displaySet
+    while (displaySetOptions.length < displaySets.length) {
+      displaySetOptions.push({});
     }
 
-    // check if their imageIds are the same
-    if (foundDisplaySet.images?.length) {
-      for (let j = 0; j < foundDisplaySet.images.length; j++) {
-        if (foundDisplaySet.images[j].imageId !== prevDisplaySet.images[j].imageId) {
-          return false;
-        }
+    // Since we only have support for dynamic data in volume viewports, we should
+    // handle this case here and set the viewportType to volume if any of the
+    // displaySets are dynamic volumes
+    viewportOptions.viewportType = displaySets.some(
```

Comment:
Can you encapsulate this into a function which can be replaced in the future?  Something which can handle WSI, video etc in the future.

Reply:
What about letting the display set itself indicate the allowed viewport types.  That way the sop class handler can define exactly what is permitted, since it actually knows that information.

Reply:
Added issue https://github.com/OHIF/Viewers/issues/4448